cmd/tsbs_generate_queries/databases/mongo/devops-naive.go

```diff
@@ -102,61 +93,263 @@ func (d *NaiveDevops) GroupByTimeAndPrimaryTag(qi query.Query, numMetrics int) {
+		matchMap := match["$match"].(bson.M)
+		matchMap["tags.hostname"] = bson.M{"$in": hostnames}
+	}
+	pipelineQuery = append(pipelineQuery, match)
+
+	project := bson.M{
+		"$project": bson.M{
+			"_id":  0,
+			"time": 1,
+			"tags": "$tags.hostname",
+		},
+	}
+	projectMap := project["$project"].(bson.M)
+	for _, metric := range metrics {
+		projectMap[metric] = 1
+	}
+	pipelineQuery = append(pipelineQuery, project)
+	pipelineQuery = append(pipelineQuery, bson.M{"$match": bson.M{"usage_user": bson.M{"$gt": 90.0}}})
+
+	humanLabel, err := devops.GetHighCPULabel("Mongo", nHosts)
+	panicIfErr(err)
 	q := qi.(*query.Mongo)
 	q.HumanLabel = []byte(humanLabel)
 	q.BsonDoc = pipelineQuery
 	q.CollectionName = []byte("point_data")
 	q.HumanDescription = []byte(fmt.Sprintf("%s: %s (%s)", humanLabel, interval.StartString(), q.CollectionName))
 }
+
+// LastPointPerHost finds the last row for every host in the dataset
+func (d *NaiveDevops) LastPointPerHost(qi query.Query) {
+	metrics := devops.GetAllCPUMetrics()
+
+	pipelineQuery := []bson.M{
+		{"$match": bson.M{"measurement": "cpu"}},
+		{
+			"$group": bson.M{
+				"_id":       bson.M{"hostname": "$tags.hostname"},
+				"last_time": bson.M{"$max": "$time"},
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id":   bson.M{"last_time": "$last_time"},
+				"hosts": bson.M{"$addToSet": "$_id.hostname"},
+			},
+		},
+		{
+			"$lookup": bson.M{
+				"from": "point_data",
+				"let":  bson.M{"time": "$_id.last_time", "hosts": "$hosts"},
+				"pipeline": []bson.M{
+					{
+						"$match": bson.M{
+							"$expr": bson.M{
+								"$and": []bson.M{
+									{"$eq": []interface{}{"$time", "$$time"}},
+									{"$in": []interface{}{"$tags.hostname", "$$hosts"}},
+									{"$eq": []interface{}{"$measurement", "cpu"}},
+								},
+							},
+						},
+					},
+					{
+						"$project": bson.M{
+							"time":          1,
+							"tags.hostname": 1,
+							"_id":           0,
+						},
+					},
+				},
+				"as": "results",
+			},
+		},
+		{
+			"$unwind": "$results",
+		},
+		{
+			"$project": bson.M{
+				"hostname":    "$results.tags.hostname",
+				"result.time": "$results.time",
+				"_id":         0,
+			},
+		},
+	}
+
+	lookupPipeline := pipelineQuery[3]["$lookup"].(bson.M)["pipeline"]
+	lookupProjectMap := lookupPipeline.([]bson.M)[1]["$project"].(bson.M)
+	projectMap := pipelineQuery[5]["$project"].(bson.M)
+	for _, metric := range metrics {
+		lookupProjectMap[metric] = 1
+		projectMap["result."+metric] = "$results." + metric
+	}
+
+	humanLabel := "Mongo last row per host"
+	q := qi.(*query.Mongo)
+	q.HumanLabel = []byte(humanLabel)
+	q.BsonDoc = pipelineQuery
+	q.CollectionName = []byte("point_data")
+	q.HumanDescription = []byte(fmt.Sprintf("%s", humanLabel))
+}
+
+// GroupByOrderByLimit populates a query.Query that has a time WHERE clause, that groups by a
+// truncated date, orders by that date, and takes a limit, e.g. in pseudo-SQL:
+//
+// SELECT date_trunc('minute', time) AS t, MAX(cpu) FROM cpu
```

Comment:
Here's an explicit date truncation in the SQL but we have them in some of the other queries in the file where they don't appear in the corresponding SQL. Are you sure the other databases are running the function every time we do?

Should we fix the SQL?

Reply:
I think there's a question here of whether the other competing systems are splitting out and storing 'minute' as a separate (redundant) column from 'time', or if 'minute' is secretly shorthand for `date_trunc('minute', time)` everywhere in this file.